src/audio/detect_test.c

```diff
@@ -0,0 +1,224 @@
+	memcpy(keyword, ipc_keyword, sizeof(struct sof_ipc_comp_process));
+
+	kd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*kd));
+	if (!kd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, kd);
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void keyword_free(struct comp_dev *dev)
+{
+	struct keyword_data *kd = comp_get_drvdata(dev);
+
+	trace_keyword("keyword_free()");
+
+	rfree(kd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int keyword_params(struct comp_dev *dev)
+{
+
+	return 0;
+}
+
+/**
+ * \brief Sets keyword control command.
+ * \param[in,out] dev keyword base component device.
+ * \param[in,out] cdata Control command data.
+ * \return Error code.
+ */
+static int keyword_ctrl_set_cmd(struct comp_dev *dev,
+			       struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int j;
+
+	/* validate */
+	if (cdata->num_elems == 0 || cdata->num_elems > SOF_IPC_MAX_CHANNELS) {
+		trace_keyword_error("keyworkd_ctrl_set_cmd() error: "
+				   "invalid cdata->num_elems");
+		return -EINVAL;
+	}
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		/*FIXME: fix logger for keywork */
+		trace_keyword("keyword_ctrl_set_cmd(), SOF_CTRL_CMD_SWITCH, ");
+
+		/* save switch state */
+		for (j = 0; j < cdata->num_elems; j++)
+			cd->switch_state[j] = cdata->chanv[j].value;
+
+		break;
+
+	default:
+		trace_keyword_error("keyword_ctrl_set_cmd() error: "
+				   "invalid cdata->cmd");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * \brief Gets keyword control command.
+ * \param[in,out] dev keyword base component device.
+ * \param[in,out] cdata Control command data.
+ * \return Error code.
+ */
+static int keyword_ctrl_get_cmd(struct comp_dev *dev,
+			       struct sof_ipc_ctrl_data *cdata, int size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int j;
+
+	/* validate */
+	if (cdata->num_elems == 0 || cdata->num_elems > SOF_IPC_MAX_CHANNELS) {
+		trace_keyword_error("invalid cdata->num_elems");
+		return -EINVAL;
+	}
+
+	if (cdata->cmd ==  SOF_CTRL_CMD_SWITCH) {
+		trace_keyword("keyword_ctrl_get_cmd(), SOF_CTRL_CMD_SWITCH");
+
+		/* copy current switch state */
+		for (j = 0; j < cdata->num_elems; j++) {
+			cdata->chanv[j].channel = j;
+			cdata->chanv[j].value = cd->switch_state[j];
+		}
+	} else {
+		trace_keyword_error("invalid cdata->cmd");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int keyword_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+
+	trace_keyword("keyword_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_VALUE:
+		return keyword_ctrl_set_cmd(dev, cdata);
+	case COMP_CMD_GET_VALUE:
+		return keyword_ctrl_get_cmd(dev, cdata, max_data_size);
+	default:
+		return -EINVAL;
+	}
+}
+
+/* copy and process stream data from source to sink buffers */
+static int keyword_copy(struct comp_dev *dev)
```

Comment:
I assume that this is an initial implementation and later on it will be expanded with pattern detection and artificial workload with parametrized characteristics  (@mluckix)  , large parameters support (@juimonen )